### ISSUE TITLE
oxlint,oxfmt: nested `.gitignore` with bare directory pattern not honored when running from inside the ignored tree

### ISSUE THREAD

**Issue by vincentllm** [oxc-project/oxc#25107]
### What version of Oxlint are you using?

1.72.0

### What command did you run?

`oxlint` (no args), executed from inside a git-ignored package directory.

### What does your `.gitignore` config look like?

A **nested** `.gitignore` (not at the repo root) ignoring a directory with the **bare directory form**:

```
# sub/.gitignore
generated
```

### What happened?

With a nested `.gitignore` using the bare directory form, running `oxlint` from inside the ignored tree still lints the files:

```bash
mkdir repro && cd repro && git init -q .
mkdir -p sub/generated/pkg
echo 'export const answer: number = "not a number";' > sub/generated/pkg/index.ts
echo 'generated' > sub/.gitignore          # nested .gitignore, bare directory form

cd sub/generated/pkg
oxlint
# => Found 0 warnings and 0 errors. Finished ... on 1 file   ← file is linted
```

Expected: `No files found to lint` — `git check-ignore -v` matches the file (git treats the whole tree as ignored).

Changing `sub/.gitignore` to the glob form fixes it:

```
generated/**
```

```
oxlint
# => No files found to lint   ← correct
```

Additional data points:

- `.gitignore` at the repository **root**: both `generated` and `generated/**` are honored correctly.
- The deviation only appears with a **nested** `.gitignore` + **bare directory** pattern + **cwd inside the ignored tree**.

Looks like the same class of issue as #22034 (which fixed the Oxfmt side).

### Environment

- oxlint 1.72.0
- Linux (x86_64)

**Comment by leaysgur** [comment]
Thanks for the report.
I changed the title because this applies to both Oxlint and Oxfmt.

What happening is:

- We use the `ignore` crate for directory walking
- The `ignore` crate respects `.gitignore` files and also handles nested ignores
- However, it makes these checks at the moment it decides whether to descend into a directory during the walk
- As a result, there are cases where it doesn't work:
  - if the CWD is already set to be ignored
  - if `oxlint target_file.ts target_dir/` is already ignored
- `rg` command (which uses `ignore` crate) also behaves the same

**Comment by jong-kyung** [comment]
@leaysgur I’d like to work on this if it’s available 👀

**Comment by leaysgur** [comment]
Oh, sorry, I'd already started on it...! 🙊 (If you are happy, let me ask you to review then?)

**Comment by jong-kyung** [comment]
> Oh, sorry, I'd already started on it...! 🙊 (If you are happy, let me ask you to review then?)

Sure! I'd be happy to review it 💜
